sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -65,35 +89,102 @@ class EquivalentExpressions {
   }
 
   /**
-   * Adds the expression to this data structure recursively. Stops if a matching expression
-   * is found. That is, if `expr` has already been added, its children are not added.
+   * Adds each expression to structural expression data structure, grouping them with existing
+   * structurally equivalent expressions. Non-recursive.
+   */
+  def addStructExpr(ctx: CodegenContext, expr: Expression): Unit = {
+    if (expr.deterministic) {
+      val refs = expr.collect {
+        case b: BoundReference => b
+      }
+
+      // For structural equivalent expressions, we need to pass in int type ordinals into
+      // split functions. If the number of ordinals is more than JVM function limit, we skip
+      // this expression.
+      // We calculate function parameter length by the number of ints plus `INPUT_ROW` plus
+      // a int type result array index.
+      val parameterLength = CodeGenerator.calculateParamLength(refs.map(_ => Literal(0))) + 2
+      if (CodeGenerator.isValidParamLength(parameterLength)) {
```

Comment:
If the length goes over the limit, the current logic gives up eliminating common exprs? If so, can we fall back into the non-structural mode?

Reply:
Good idea.